### PR TITLE
Clarify server HRR behavior, and add label-based workaround for nonce-reuse

### PR DESCRIPTION
**Pull request by chris-wood**
Since the client's ESNI key share doesn't change across CHs in response to a HRR, the same AEAD IV would be used to encrypt/decrypt two different messages. This is an attempt to fix that, while also specifying the server-side HRR behavior. (Basically, check that the ESNI contents have not changed.)

### CODE REVIEW

File: draft-ietf-tls-esni.md [hunk 1]
```diff
@@ -772,6 +778,11 @@ the TLS connection to the backend server (if in Split Mode). In
 the latter case, it does not make any changes to the TLS
 messages, but just blindly forwards them.
 
+If the ClientHello is the result of a HelloRetryRequest, servers MUST
+abort the connection with an "illegal_parameter" alert if the decrypted
+ClientESNIInner from the second ClientHello does not match that
+of the first ClientHello.
```
Comment (kazuho) [inline draft-ietf-tls-esni.md]: I agree that servers MUST verify the value of the ESNI extension, but the fields being required to be validated seems to differ from the the fields that are required to be consistent. Is there a reason for the two to be different?

FWIW, "Handling the server response" section states as follows:
> If the server sends a HelloRetryRequest in response to the ClientHello and the client can send a second updated ClientHello per the rules in {{RFC8446}}, the "encrypted_server_name" extension values which do not depend on the (possibly updated) KeyShareClientHello, i.e,, ClientEncryptedSNI.suite, ClientEncryptedSNI.key_share, and ClientEncryptedSNI.record_digest, MUST NOT change across ClientHello messages. Moreover, ClientESNIInner.nonce and ClientESNIInner.realSNI MUST not change across ClientHello messages. Informally, the values of all unencrypted extension information, as well as the inner extension plaintext, must be consistent between the first and second ClientHello messages.

Comment (chris-wood) [reply draft-ietf-tls-esni.md]: No reason -- just an oversight on my part. It should be fixed now!

**Comment by ekr** [comment]
Nice catch, Would it make more sense to instead use the KeyShare as a salt in the HKDF?

**Comment by martinthomson** [comment]
Or you could include the handshake transcript up to, but not including, the ClientHello.  That is the empty string before HRR< but something else after.

**Comment by davidben** [comment]
> Nice catch, Would it make more sense to instead use the KeyShare as a salt in the HKDF?

That is a little iffy with cookie-only HelloRetryRequests. Those don't change the ClientHello key shares list, so the client would derive the same key + nonce pair. The client would also use the same plaintext so it's fine, but it seems an unnecessarily fragility.

**Comment by ekr** [comment]
Agreed. I think I prefer the "new DH" version.

**Pull request merged**